featuregates/controller/pkg/feature/feature_controller.go

```diff
@@ -0,0 +1,242 @@
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// reconcileFeatureInFeatureGateSpec reconciles feature that is present in FeatureGate spec
+func reconcileFeatureInFeatureGateSpec(ctx context.Context, c client.Client, featureGate *corev1alpha2.FeatureGate, feature *corev1alpha2.Feature) error {
+	policy := corev1alpha2.GetPolicyForStabilityLevel(feature.Spec.Stability)
+	featureReference, _ := util.GetFeatureReferenceFromFeatureGate(featureGate, feature.Name)
+	featureResult, activate := applyPolicyToComputeFeatureResultAndActivation(policy, featureReference, feature.Spec.Stability)
+
+	// Update FeatureGate status
+	featureGate.Status.Results = computeFeatureGateStatusResults(featureGate.Status, featureResult, true)
+	if err := c.Status().Update(ctx, featureGate); err != nil {
+		return fmt.Errorf("couldn't update %s FeatureGate status :%w", featureGate.Name, err)
+	}
+
+	// Update Feature status
+	feature.Status.Activated = activate
+	if err := c.Update(ctx, feature); err != nil {
+		return fmt.Errorf("couldn't update %s Feature status :%w", feature.Name, err)
+	}
+	return nil
+}
+
+// reconcileDeletedFeature reconciles feature that has been deleted
+func reconcileDeletedFeature(ctx context.Context, c client.Client, featureName string) error {
+	// Check if feature part of any FeatureGate spec and update status in Feature result to Invalid
+	featureGate, found, err := util.GetFeatureGateForFeature(ctx, c, featureName)
+	if err != nil {
+		return err
+	}
+	if found {
+		featureGate.Status.Results = computeFeatureGateStatusResults(featureGate.Status, corev1alpha2.Result{
+			Name:    featureName,
+			Status:  constants.InvalidFeatureReferenceStatus,
+			Message: "Feature doesn't exist in cluster",
+		}, true)
+		if err := c.Status().Update(ctx, featureGate); err != nil {
+			return fmt.Errorf("couldn't update %s FeatureGate status :%w", featureGate.Name, err)
+		}
+	}
+	return nil
+}
+
+// reconcileFeatureNotInFeatureGateSpec reconciles feature that is not found in any FeatureGate spec
+func reconcileFeatureNotInFeatureGateSpec(ctx context.Context, c client.Client, feature *corev1alpha2.Feature) error {
+	// If feature is not part of any FeatureGate spec, check for its existence in the FeatureGate status.
+	// If found in the FeatureGate status, remove its entry from status and update the feature status to deactivated
+	featureGate, found, err := util.GetFeatureGateWithFeatureInStatus(ctx, c, feature.Name)
+	if err != nil {
+		return err
+	}
+	if found {
+		// Remove feature from FeatureGate status
+		featureGate.Status.Results = computeFeatureGateStatusResults(featureGate.Status, corev1alpha2.Result{
+			Name: feature.Name,
+		}, false)
+		if err := c.Status().Update(ctx, featureGate); err != nil {
+			return fmt.Errorf("couldn't update %s FeatureGate status :%w", featureGate.Name, err)
+		}
+	}
+	// Update Feature status to set feature as deactivated
+	feature.Status.Activated = false
```

Comment:
If the feature is no longer referenced from any feature gate, shouldn't we set the feature to its default activation state? (Which might be true in some cases!)